state-chain/pallets/cf-ingress-egress/src/lib.rs

```diff
@@ -3038,32 +3093,49 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		let mut deposit_channel =
+			match PreallocatedChannels::<T, I>::mutate(requester, |queue| queue.pop_front()) {
+				Some(channel) => channel,
+				// if there are no pre-allocated channels, take the next channel
+				// from the pool of recycled channels, or generate a new channel entirely.
+				None => match DepositChannelPool::<T, I>::drain().next() {
+					Some((_, deposit_channel)) => deposit_channel,
+					None => Self::generate_new_channel(source_asset)?,
+				},
+			};
+		// Make sure to set the asset on the channel, in case it was recycled
+		deposit_channel.asset = source_asset;
+
+		// Proactively pre-allocate new channels for the requester,
+		// This is done after the above loop deliberately to ensure that we always respect per-role
+		// preallocated channels limits and cater for cases when the configured
+		// MaximumPreallocatedChannels changes, or a channel is allocated from the pre-allocated
+		// channels list.
+		for _ in 0..(MaximumPreallocatedChannels::<T, I>::get(T::AccountRoleRegistry::account_role(
+			requester,
+		)) as usize)
+			.saturating_sub(PreallocatedChannels::<T, I>::get(requester).len())
 		{
-			deposit_channel.asset = source_asset;
-			(deposit_channel, channel_id)
-		} else {
-			let next_channel_id = Self::allocate_next_channel_id()?;
-			(
-				DepositChannel::generate_new::<T::AddressDerivation>(next_channel_id, source_asset)
-					.map_err(|e| match e {
-						AddressDerivationError::MissingPolkadotVault =>
-							Error::<T, I>::MissingPolkadotVault,
-						AddressDerivationError::MissingBitcoinVault =>
-							Error::<T, I>::MissingBitcoinVault,
-						AddressDerivationError::BitcoinChannelIdTooLarge =>
-							Error::<T, I>::BitcoinChannelIdTooLarge,
-						AddressDerivationError::SolanaDerivationError { .. } =>
-							Error::<T, I>::SolanaAddressDerivationError,
-						AddressDerivationError::MissingSolanaApiEnvironment =>
-							Error::<T, I>::MissingSolanaApiEnvironment,
-						AddressDerivationError::MissingAssethubVault =>
-							Error::<T, I>::MissingAssethubVault,
-					})?,
-				next_channel_id,
-			)
-		};
+			// For chains that preallocation is not enabled such as Ethereum and Arbitrum,
+			// only pre-allocate channels from the global pool of addresses, for other chains,
+			// preallocate a newly generated channel.
+			if !Into::<ForeignChain>::into(source_asset).can_preallocate_addresses() {
+				if let Some((_, reused_channel)) = DepositChannelPool::<T, I>::drain().next() {
+					PreallocatedChannels::<T, I>::mutate(requester, |queue| {
+						queue.push_back(reused_channel)
+					});
+				}
+			} else {
+				// It doesn't matter what source asset we pass here, because source_asset
+				// is ignored by chains that support preallocation.
+				let new_channel = Self::generate_new_channel(source_asset)?;
+				PreallocatedChannels::<T, I>::mutate(requester, |queue| {
+					queue.push_back(new_channel)
+				});
+			}
+		}
 
 		let deposit_address = deposit_channel.address.clone();
```

Comment:
don't actually need this separate clone it seems (and would just remove the variable in that case)